### ISSUE TITLE
loading data without dataiter

### ISSUE THREAD

**Issue by GeorgeXia1828** [apache/incubator-mxnet#1830]
I am trying to loading data batch by batch,  I found the grads is not right, much bigger? Is anything wrong with the code? Or, is there any example train like this way? As it do in lasagne.

import mxnet as mx
import numpy as np
import struct
import time
## create the network

def net():
    data = mx.symbol.Variable('data')
    fc1 = mx.symbol.FullyConnected(data, name='fc1', num_hidden=128)
    act1 = mx.symbol.Activation(fc1, name='act1', act_type='relu')
    fc2 = mx.symbol.FullyConnected(act1, name='fc2', num_hidden=128)
    act2 = mx.symbol.Activation(fc2, name='act2', act_type='relu')
    fc3 = mx.symbol.FullyConnected(act2, name='fc3', num_hidden=10)
    softmax = mx.symbol.SoftmaxOutput(fc3, name='sm')
    return softmax

def generate_img(path):
    binfile = open(path)
    buf = binfile.read()
    index = struct.calcsize('>IIII')
    img = struct.unpack_from('>' + str(50000_784) + 'B', buf, index)
    img = np.array(img)
    train = img.reshape(-1, 784)
    index += struct.calcsize('>' + str(50000_784) + 'B')

```
img = struct.unpack_from('>' + str(10000*784) + 'B', buf, index)
img = np.array(img)
val = img.reshape(-1, 784)
return train/np.float32(256), val/np.float32(256)
```

def generate_label(path):
    binfile = open(path)
    buf = binfile.read()
    index = struct.calcsize('>II')
    label = struct.unpack_from('>50000B', buf, index)
    train = np.array(label)
    index += struct.calcsize('>50000B')

```
label = struct.unpack_from('>10000B', buf, index)
val = np.array(label)
return train/np.float32(256), val/np.float32(256)
```

if **name** == '**main**':
    ##load the data
    train_data = '../data/train-images-idx3-ubyte'
    train_label = '../data/train-labels-idx1-ubyte'
    test_data = '../data/t10k-images-idx3-ubyte'
    test_label = '../data/t10k-labels-idx1-ubyte'

```
#prepare data
train, val = generate_img(train_data)
train_label, val_label = generate_label(train_label)

softmax = net()

##prepare data
data = mx.nd.array(train)
fc1_weight_0 = np.random.uniform(-0.1, 0.1, (128, 784))
fc1_bias_0 = np.zeros((128,))
fc2_weight_0 = np.random.uniform(-0.1, 0.1, (128, 128))
fc2_bias_0 = np.zeros((128,))
fc3_weight_0 = np.random.uniform(-0.1, 0.1, (10, 128))
fc3_bias_0 = np.zeros((10,))

    data_1 = np.zeros((50000, 784))
fc1_weight_1 = np.zeros((128, 784))
fc1_bias_1 = np.zeros((128,))
fc2_weight_1 = np.zeros((128, 128))
fc2_bias_1 = np.zeros((128,))
fc3_weight_1 = np.zeros((10, 128))
fc3_bias_1 = np.zeros((10,))
sm_label_1 = np.zeros((50000,))

fc1_weight_0 = mx.nd.array(fc1_weight_0)
fc1_bias_0 = mx.nd.array(fc1_bias_0)
fc2_weight_0 = mx.nd.array(fc2_weight_0)
fc2_bias_0 = mx.nd.array(fc2_bias_0)
fc3_weight_0 = mx.nd.array(fc3_weight_0)
fc3_bias_0 = mx.nd.array(fc3_bias_0)
sm_label_0 = mx.nd.array(train_label)

    data_1 = mx.nd.array(data_1)
fc1_weight_1 = mx.nd.array(fc1_weight_1)
fc1_bias_1 = mx.nd.array(fc1_bias_1)
fc2_weight_1 = mx.nd.array(fc2_weight_1)
fc2_bias_1 = mx.nd.array(fc2_bias_1)
fc3_weight_1 = mx.nd.array(fc3_weight_1)
fc3_bias_1 = mx.nd.array(fc3_bias_1)
sm_label_1 = mx.nd.array(sm_label_1)

args = {'data':data, 'fc1_weight':fc1_weight_0, 'fc1_bias':fc1_bias_0, 'fc2_weight':fc2_weight_0, 'fc2_bias':fc2_bias_0,
        'fc3_weight':fc3_weight_0, 'fc3_bias':fc3_bias_0, 'sm_label':sm_label_0}

    args_grad = {'data':data_1, 'fc1_weight':fc1_weight_1, 'fc1_bias':fc1_bias_1, 'fc2_weight':fc2_weight_1, 'fc2_bias':fc2_bias_1,
        'fc3_weight':fc3_weight_1, 'fc3_bias':fc3_bias_1, 'sm_label':sm_label_1}
#print args_grad['fc1_bias'].asnumpy()

    reqs = {'data':'null', 'fc1_weight':'write', 'fc1_bias':'write', 'fc2_weight':'write', 'fc2_bias':'write',
        'fc3_weight':'write', 'fc3_bias':'write', 'sm_label':'null'}

##training the model
exe = softmax.bind(ctx=mx.context.cpu(), args=args, args_grad=args_grad, grad_req=reqs)
opt = mx.optimizer.SGD(learning_rate=0.01)
start_time = time.time()

exe.forward()
exe.backward()
opt.update(index=1, weight=fc1_bias_0, grad=fc1_bias_1, state=None)

mx.nd.waitall()
print args_grad['fc1_bias'].asnumpy()
```

**Comment by pluskid** [comment]
One possible reason is that by default, when constructing the optimizer, we further divide the learning rate by mini-batch size. So that might be why your updates looks "bigger".

BTW: with the new module API, it is now less complicated to manually create a data batch and then call forward / backward / update manually. See [the example](https://github.com/dmlc/mxnet/blob/master/example/module/mnist_mlp.py#L30) here.

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
